File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        unordered_map<char, int> mp;
        
        int n = s.length(), start = 0, maxi = INT_MIN;

        for (int end = 0; end < n; end++) {
            mp[s[end]]++;
            
            while(mp.size() != end+1 - start){
                mp[s[start]]--;
                
                if (mp[s[start]] == 0) mp.erase(s[start]);
                start++;
            }
            
            maxi = max(maxi, end+1 - start);
        }
        
        return maxi == INT_MIN?0:maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer> map = new HashMap<>();
        
        int i = 0;
        int j = 0;
        int maxLength = 0;
        while(j<s.length()){
            
            if(map.containsKey(s.charAt(j))){
                
                i = Math.max(i,map.get(s.charAt(j))+1);
                
            }
            map.put(s.charAt(j),j);
            maxLength = Math.max(maxLength,j-i+1);
            j++;
        }
        
        return maxLength;
    }
}